src/libraries/System.Data.Common/src/System/Data/Common/DbDataSource.cs

```diff
@@ -0,0 +1,570 @@
+                    }
+                    catch (Exception e)
+                    {
+                        ExceptionBuilder.TraceExceptionWithoutRethrow(e);
+
+                        // Swallow to allow the original exception to bubble up
+                        // Also, refrain from bubbling up the close exception even if there's no original exception,
+                        // since it's not relevant to the user - execution did complete successfully, and the connection
+                        // close is just an internal detail that shouldn't cause user code to fail.
+                    }
+                }
+            }
+
+            protected override DbDataReader ExecuteDbDataReader(CommandBehavior behavior)
+            {
+                _connection.Open();
+
+                try
+                {
+                    return _wrappedBatch.ExecuteReader(behavior | CommandBehavior.CloseConnection);
+                }
+                catch
+                {
+                    try
+                    {
+                        _connection.Close();
+                    }
+                    catch (Exception e)
+                    {
+                        ExceptionBuilder.TraceExceptionWithoutRethrow(e);
+
+                        // Swallow to allow the original exception to bubble up
+                    }
+
+                    throw;
+                }
+            }
+
+            protected override async Task<DbDataReader> ExecuteDbDataReaderAsync(
+                CommandBehavior behavior,
+                CancellationToken cancellationToken)
+            {
+                await _connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+
+                try
+                {
+                    return await _wrappedBatch.ExecuteReaderAsync(
+                            behavior | CommandBehavior.CloseConnection,
+                            cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                catch
+                {
+                    try
+                    {
+                        await _connection.CloseAsync().ConfigureAwait(false);
+                    }
+                    catch (Exception e)
+                    {
+                        ExceptionBuilder.TraceExceptionWithoutRethrow(e);
+
+                        // Swallow to allow the original exception to bubble up
+                    }
+
+                    throw;
+                }
+            }
+
+            protected override DbBatchCommand CreateDbBatchCommand() => throw new NotImplementedException();
+
+            public override void Cancel()
+                => _wrappedBatch.Cancel();
+
+            public override void Prepare()
+                => _wrappedBatch.Prepare();
+
+            public override Task PrepareAsync(CancellationToken cancellationToken = default) => throw new NotImplementedException();
```

Comment:
```suggestion
            public override Task PrepareAsync(CancellationToken cancellationToken = default)
                => await _wrappedBatch.PrepareAsync(cancellationToken);
```

Either both of Prepare/PrepareAsync should be implemented, or neither.

Reply:
Yep, just an oversight, thanks.